.github/workflows/changelog.yml

```diff
@@ -1,31 +1,68 @@
+          echo "TAGGED_MILESTONE=${{ github.event.pull_request.milestone.title }}" >> $GITHUB_ENV
       - uses: actions/checkout@v3
-      - name: Check CHANGES.rst Updated
+        with:
+          fetch-depth: '0'
+      - name: Check the changelog entry
         run: |
-          git fetch origin main:main
-          updated_files=$(git diff main -- CHANGES.rst)
-          if [[ -z $updated_files ]]
+          set -xe
+          changed_files=$(git diff --name-only origin/main)
+          # Changelog should be updated only if tests have been modified
+          if [[ ! "$changed_files" =~ tests ]]
           then
-            exit 1
+            exit 0
           fi
-
-      - name: Comment PR
-        if: ${{ failure() }}
-        uses: thollander/actions-comment-pull-request@v1
-        with:
-          comment_includes: "changelog" # To avoid creating additional comments
-          message: |
-            Please include the change for this PR in the changelog (`CHANGES.rst`).
-          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          all_changelogs=$(cat CHANGES.rst)
+          if [[ "$all_changelogs" =~ :pr:\`$PR_NUMBER\` ]]
+          then
+            echo "Changelog has been updated."
+            # If the pull request is milestoned check the correspondent changelog
+            if exist -f CHANGES.rst${TAGGED_MILESTONE:0:4}.rst
+            then
+              expected_changelog=$(cat CHANGES.rst${TAGGED_MILESTONE:0:4}.rst)
+              if [[ "$expected_changelog" =~ :pr:\`$PR_NUMBER\` ]]
+              then
+                echo "Changelog and milestone correspond."
+              else
+                echo "Changelog and milestone do not correspond."
+                echo "If you see this error make sure that the tagged milestone for the PR"
+                echo "and the edited changelog filename properly match."
+                exit 1
+              fi
+            fi
+          else
+            echo "A Changelog entry is missing."
+            echo ""
+            echo "Please add an entry to the changelog at 'CHANGES.rst'"
+            echo "to document your change assuming that the PR will be merged"
+            echo "in time for the next release of dirty-cat."
+            echo ""
+            echo "Look at other entries in that file for inspiration and please"
+            echo "reference this pull request using the ':pr:' directive and"
+            echo "credit yourself (and other contributors if applicable) with"
+            echo "the ':user:' directive."
+            echo ""
+            echo "If you see this error and there is already a changelog entry,"
+            echo "check that the PR number is correct."
+            echo ""
+            echo "If you believe that this PR does not warrant a changelog"
+            echo "entry, say so in a comment so that a maintainer will label"
+            echo "the PR with 'No Changelog Needed' to bypass this check."
```

Comment:
And this one seems to indicate that there's a github label dedicated! I guess we can create it, we just have to check if there's logic linked to it.

Edit: logic there is! Line 13
```bash
if: ${{ contains(github.event.pull_request.labels.*.name, 'No Changelog Needed') == 0 }}
```